File: java-samples/corda5-obligation-cordapp/contracts/src/main/java/com/r3/developers/samples/obligation/contracts/IOUContract.java
package com.r3.developers.samples.obligation.contracts;

import com.r3.developers.samples.obligation.states.IOUState;
import net.corda.v5.base.exceptions.CordaRuntimeException;
import net.corda.v5.ledger.utxo.Command;
import net.corda.v5.ledger.utxo.Contract;
import net.corda.v5.ledger.utxo.ContractState;
import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction;

import java.security.PublicKey;
import java.util.Set;


public class IOUContract implements Contract {


    @Override
    public boolean isRelevant(ContractState state, Set<PublicKey> myKeys) {
        return Contract.super.isRelevant(state, myKeys);
    }

    public static class Issue implements Command { }
    public static class Settle implements Command { }
    public static class Transfer implements Command { }

    @Override
    public void verify(UtxoLedgerTransaction transaction) {

        requireThat( transaction.getCommands().size() == 1, "Require a single command.");
        Command command = transaction.getCommands().get(0);
        IOUState output = transaction.getOutputStates(IOUState.class).get(0);
        requireThat(output.getParticipants().size() == 2, "The output state should have two and only two participants.");

        if(command.getClass() == IOUContract.Issue.class) {
            requireThat(transaction.getOutputContractStates().size() == 1, "Only one output states should be created when issuing an IOU.");

        }else if(command.getClass() == IOUContract.Transfer.class) {
            requireThat( transaction.getInputContractStates().size() > 0, "There must be one input IOU.");
        }
        else if(command.getClass() == IOUContract.Settle.class) {
            requireThat( transaction.getInputContractStates().size() > 0, "There must be one input IOU.");
        }
        else {
            throw new CordaRuntimeException("Unsupported command");
        }
    }
    private void requireThat(boolean asserted, String errorMessage) {
        if(!asserted) {
            throw new CordaRuntimeException("Failed requirement: " + errorMessage);
        }
    }

}

File: kotlin-samples/corda5-obligation-cordapp/contracts/src/main/kotlin/com/r3/developers/samples/obligation/contracts/IOUContract.kt
package com.r3.developers.samples.obligation.contracts

import com.r3.developers.samples.obligation.states.IOUState
import net.corda.v5.base.exceptions.CordaRuntimeException
import net.corda.v5.ledger.utxo.Command
import net.corda.v5.ledger.utxo.Contract
import net.corda.v5.ledger.utxo.transaction.UtxoLedgerTransaction

class IOUContract : Contract {

    //IOU Commands
    class Issue: Command
    class Settle: Command
    class Transfer: Command

    override fun verify(transaction: UtxoLedgerTransaction) {

        // Ensures that there is only one command in the transaction
        val command = transaction.commands.singleOrNull() ?: throw CordaRuntimeException("Requires a single command.")

        "The output state should have two and only two participants." using {
            val output = transaction.outputContractStates.first() as IOUState
            output.participants.size== 2
        }
        // Switches case based on the command
        when(command) {
            // Rules applied only to transactions with the Issue Command.
            is Issue -> {
                "When command is Create there should be one and only one output state." using (transaction.outputContractStates.size == 1)
            }
            // Rules applied only to transactions with the Settle Command.
            is Settle -> {
                "When command is Update there should be one and only one output state." using (transaction.outputContractStates.size == 1)
            }
            // Rules applied only to transactions with the Transfer Command.
            is Transfer -> {
                "When command is Update there should be one and only one output state." using (transaction.outputContractStates.size == 1)
            }
            else -> {
                throw CordaRuntimeException("Command not allowed.")
            }
        }
    }

    // Helper function to allow writing constraints in the Corda 4 '"text" using (boolean)' style
    private infix fun String.using(expr: Boolean) {
        if (!expr) throw CordaRuntimeException("Failed requirement: $this")
    }

    // Helper function to allow writing constraints in '"text" using {lambda}' style where the last expression
    // in the lambda is a boolean.
    private infix fun String.using(expr: () -> Boolean) {
        if (!expr.invoke()) throw CordaRuntimeException("Failed requirement: $this")
    }
}